impacted-test-engine/src/main/java/com/teamscale/test_impacted/test_descriptor/TestDescriptorUtils.java

```diff
@@ -0,0 +1,133 @@
+	 * test cases.
+	 */
+	public static boolean isTestRepresentative(TestDescriptor testDescriptor) {
+		boolean isTestTemplateOrTestFactory = isTestTemplateOrTestFactory(testDescriptor);
+		boolean isNonParameterizedTest = testDescriptor.isTest() && !isTestTemplateOrTestFactory(
+				testDescriptor.getParent().get());
+		return isNonParameterizedTest || isTestTemplateOrTestFactory;
+	}
+
+	/**
+	 * Returns true if a {@link TestDescriptor} represents a test template or a test factory.
+	 * <p>
+	 * An example of a {@link UniqueId} of the {@link TestDescriptor} is:
+	 * <p>
+	 * {@code [engine:junit-jupiter]/[class:com.example.project.JUnit5Test]/[test-template:withValueSource(java.lang.String)]}
+	 */
+	private static boolean isTestTemplateOrTestFactory(TestDescriptor testDescriptor) {
+		if (testDescriptor == null) {
+			return false;
+		}
+		List<Segment> segments = testDescriptor.getUniqueId().getSegments();
+
+		if (segments.isEmpty()) {
+			return false;
+		}
+
+		String lastSegmentType = segments.get(segments.size() - 1).getType();
+		return "test-template".equals(lastSegmentType) || "test-factory".equals(lastSegmentType);
+	}
+
+	/** Creates a stream of the test representatives contained by the {@link TestDescriptor}. */
+	public static Stream<TestDescriptor> streamTestRepresentatives(TestDescriptor testDescriptor) {
+		if (isTestRepresentative(testDescriptor)) {
+			return Stream.of(testDescriptor);
+		}
+		return testDescriptor.getChildren().stream().flatMap(TestDescriptorUtils::streamTestRepresentatives);
+	}
+
+	/**
+	 * Returns the {@link Segment#getValue()} matching the type or {@link Optional#empty()} if no matching segment can
+	 * be found.
+	 */
+	public static Optional<String> getUniqueIdSegment(TestDescriptor testDescriptor, String type) {
+		return testDescriptor.getUniqueId().getSegments().stream().filter(segment -> segment.getType().equals(type))
+				.findFirst().map(
+						Segment::getValue);
+	}
+
+	/** Returns {@link TestDetails#sourcePath} for a {@link TestDescriptor}. */
+	public static String getSource(TestDescriptor testDescriptor) {
+		Optional<TestSource> source = testDescriptor.getSource();
+		if (source.isPresent() && source.get() instanceof MethodSource) {
+			MethodSource ms = (MethodSource) source.get();
+			return ms.getClassName().replace('.', '/');
+		}
+		return null;
+	}
+
+	/** Returns the {@link AvailableTests} contained within the root {@link TestDescriptor}. */
+	public static AvailableTests getAvailableTests(TestEngine testEngine, TestDescriptor rootTestDescriptor) {
+		AvailableTests availableTests = new AvailableTests();
+		ITestDescriptorResolver testDescriptorResolver = TestDescriptorResolverRegistry
+				.getTestDescriptorResolver(testEngine);
+
+		TestDescriptorUtils.streamTestRepresentatives(rootTestDescriptor)
+				.forEach(testDescriptor -> {
+					Optional<String> clusterId = testDescriptorResolver.getClusterId(testDescriptor);
```

Comment:
When I run this locally I get a NPE here:
```
 WARNING: TestEngine with ID 'teamscale-test-impacted' failed to execute tests
    java.lang.NullPointerException
        at com.teamscale.test_impacted.test_descriptor.TestDescriptorUtils.lambda$getAvailableTests$4(TestDescriptorUtils.java:110)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
        at java.util.stream.Streams$StreamBuilderImpl.forEachRemaining(Streams.java:419)
        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
        at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
        at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
        at com.teamscale.test_impacted.test_descriptor.TestDescriptorUtils.getAvailableTests(TestDescriptorUtils.java:109)
        at com.teamscale.test_impacted.engine.ImpactedTestEngine.runTestExecutor(ImpactedTestEngine.java:93)
        at com.teamscale.test_impacted.engine.ImpactedTestEngine.execute(ImpactedTestEngine.java:75)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)

```

Reply:
However the build silently passes with success

Reply:
Global renaming refactoring was too eager.